userd/autostart.go

```diff
@@ -0,0 +1,210 @@
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil/shlex"
+)
+
+// expandDesktopFields processes the input string and expands any %<char>
+// patterns. '%%' expands to '%', all other patterns expand to empty strings.
+func expandDesktopFields(in string) string {
+	raw := []rune(in)
+	out := make([]rune, 0, len(raw))
+
+	var hasKey bool
+	for _, r := range raw {
+		if hasKey {
+			hasKey = false
+			// only allow %% -> % expansion, drop other keys
+			if r == '%' {
+				out = append(out, r)
+			}
+			continue
+		} else if r == '%' {
+			hasKey = true
+			continue
+		}
+		out = append(out, r)
+	}
+	return string(out)
+}
+
+func findExec(desktopFileContent []byte) (string, error) {
+	scanner := bufio.NewScanner(bytes.NewBuffer(desktopFileContent))
+	execCmd := ""
+	for scanner.Scan() {
+		bline := scanner.Bytes()
+
+		if !bytes.HasPrefix(bline, []byte("Exec=")) {
+			continue
+		}
+
+		full := string(bline[len("Exec="):])
+		execCmd = expandDesktopFields(full)
+		break
+	}
+
+	execCmd = strings.TrimSpace(execCmd)
+	if execCmd == "" {
+		return "", fmt.Errorf("Exec not found or invalid")
+	}
+	return execCmd, nil
+}
+
+func getCurrentSnapInfo(snapName string) (*snap.Info, error) {
+	curFn := filepath.Join(dirs.SnapMountDir, snapName, "current")
+	realFn, err := os.Readlink(curFn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find current revision for snap %s: %s", snapName, err)
+	}
+	rev := filepath.Base(realFn)
+	revision, err := snap.ParseRevision(rev)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read revision %s: %s", rev, err)
+	}
+
+	return snap.ReadInfo(snapName, &snap.SideInfo{Revision: revision})
+}
+
+func autostartCmd(snapName, desktopFilePath string) (*exec.Cmd, error) {
+	desktopFile := filepath.Base(desktopFilePath)
+
+	info, err := getCurrentSnapInfo(snapName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain snap information for snap %q: %v", snapName, err)
+	}
+
+	var app *snap.AppInfo
+	for _, candidate := range info.Apps {
+		if candidate.Autostart == desktopFile {
+			app = candidate
+			break
+		}
+	}
+	if app == nil {
+		return nil, fmt.Errorf("cannot match desktop file with snap %q applications", snapName)
+	}
+
+	content, err := ioutil.ReadFile(desktopFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: Ignore all fields and just look for Exec=..., this also means
+	// that fields with meaning such as TryExec, X-GNOME-Autostart and so on
+	// are ignored
+
+	command, err := findExec(content)
+	if err != nil {
+		return nil, fmt.Errorf("cannot determine startup command: %v", err)
+	}
+	logger.Debugf("exec line: %v", command)
+
+	split, err := shlex.Split(command)
+	if err != nil {
+		return nil, fmt.Errorf("invalid application startup command: %v", err)
+	}
+
+	cmd := exec.Command(app.WrapperPath(), split[1:]...)
```

Comment:
Maybe a comment here that (or elsewhere) that the `Exec=command params...` command is ignored and replaced with the app. The code shows and the forum post but we don't document it in this PR afaict. Maybe its enough to put the (few) lines from the forum topic in this source files to give a high level overview how this works?